README.md
<div align="center">

  <h1 align="middle">🎱 로또 게임</h1>
   <img src="https://img.shields.io/badge/node-18.17.1-339933?logo=node.js">
</div>

<br>

# ❓ How To Run

## Install

- 테스트 패키지 설치를 위해 Node.js 버전 18.17.1 이상이 필요하다.
- 다음 명령어를 입력해 패키지를 설치한다.

```
npm install
```

## Run

- 다음 명령어를 입력해 프로그램을 실행한다.

```
node ./src/App.js
```

## Test

- 설치가 완료되었다면, 다음 명령어를 입력해 테스트를 실행한다.

```
npm test
```

# 폴더 구조 

```
📦src
 ┣ 📂controller
 ┃ ┗ 📜LottoController.js : 로또 게임 도메인과 view 를 이어주는 중계 역할 수행
 ┣ 📂domain 
 ┃ ┣ 📜BonusPrize.js
 ┃ ┣ 📜Customer.js
 ┃ ┣ 📜Lotto.js
 ┃ ┣ 📜LottoFactory.js
 ┃ ┣ 📜LottoWinResult.js
 ┃ ┣ 📜Prize.js
 ┃ ┗ 📜WinningLotto.js
 ┣ 📂utils
 ┣ 📂view
 ┃ ┣ 📜InputView.js
 ┃ ┗ 📜OutputView.js
 ┣ 📜App.js : 어플리케이션 실행 흐름을 관리
 ┣ 📜Constant.js : 게임 내에 사용되는 상수 관리
 ┣ 📜CustomError.js 
 ┗ 📜index.js
```

# 🛠️ 주어진 요구사항

로또 게임 기능을 구현해야 한다. 로또 게임은 아래와 같은 규칙으로 진행된다.
- 로또 번호의 숫자 범위는 1~45까지이다.
- 1개의 로또를 발행할 때 중복되지 않는 6개의 숫자를 뽑는다.
- 당첨 번호 추첨 시 중복되지 않는 숫자 6개와 보너스 번호 1개를 뽑는다.
- 당첨은 1등부터 5등까지 있다. 당첨 기준과 금액은 아래와 같다.
    - 1등: 6개 번호 일치 / 2,000,000,000원
    - 2등: 5개 번호 + 보너스 번호 일치 / 30,000,000원
    - 3등: 5개 번호 일치 / 1,500,000원
    - 4등: 4개 번호 일치 / 50,000원
    - 5등: 3개 번호 일치 / 5,000원
로또 구입 금액을 입력하면 구입 금액에 해당하는 만큼 로또를 발행해야 한다.
로또 1장의 가격은 1,000원이다.
당첨 번호와 보너스 번호를 입력받는다.
사용자가 구매한 로또 번호와 당첨 번호를 비교하여 당첨 내역 및 수익률을 출력하고 로또 게임을 종료한다.
사용자가 잘못된 값을 입력할 경우 throw문을 사용해 예외를 발생시킨다. 그런 다음, "[ERROR]"로 시작하는 에러 메시지를 출력하고 해당 부분부터 입력을 다시 받는다.
예시) [ERROR] 숫자가 잘못된 형식입니다.

# 📌 구현 기능 목록

# 기능 구현 목록

## 1. 로또 구매하기

- [x] 로또 구입 금액을 입력 받는다. ( 1,000원 단위 )

  - [x] 숫자가 아닌 경우
  - [x] 양의 정수가 아닌 경우
  - [x] 1000 단위로 나누어 떨어지지 않는 경우
  - [x] 한 장의 로또도 구매할 수 없는 경우 ( 1,000원 미만 )
  - [x] 최대 입력값 제한 : 10만원

- [x] 로또를 생성한다.

  - [x] 1 ~ 45 사이 숫자가 아닌 경우
  - [x] 당첨 번호가 중복 되는 경우
  - [x] 번호가 6개가 아닌 경우 ( = 아무런 값도 입력하지 않은 경우, 개수가 넘치는 경우 포함 )

- [x] 살 수 있는 만큼 ( 구입 금액 / 1000 ) 로또를 구매(발행)한다.
- [x] 발행한 로또 수량 및 번호를 출력한다.
  ```
  8개를 구매했습니다.
  [8, 21, 23, 41, 42, 43]
  [3, 5, 11, 16, 32, 38]
  [7, 11, 16, 35, 36, 44]
  [1, 8, 11, 31, 41, 42]
  [13, 14, 16, 38, 42, 45]
  [7, 11, 30, 40, 42, 43]
  [2, 13, 22, 32, 38, 45]
  [1, 3, 5, 14, 22, 45]
  ```

## 2. 로또 결과 생성하기

- [x] 당첨 번호를 입력받는다. ( 번호는 쉼표를 기준으로 구분한다. )
- [x] 보너스 번호를 입력받는다.
- [x] 우승 로또를 생성한다.

  - [x] 보너스가 정수가 아닌 경우
  - [x] 보너스가 1 ~ 45 사이의 숫자가 아닌 경우
  - [x] 보너스가 기존 당첨 번호들과 겹칠 경우

- [x] 로또 당첨 결과를 생성한다.

## 3. 로또 통계 처리하기

- [x] 사용자 로또 하나와 당첨 로또를 비교해서 보상을 구한다.
- [x] 구매한 전체 로또와 당첨 번호를 비교하면서 당첨 통계(보상의 종류 : 개수)을 구한다.
- [x] 총 수익률( 당첨 금액 합 / 투자한 비용 ) \* 100 % 을 계산한다.
  - 수익률은 소수점 둘째 자리에서 반올림한다.
- [x] 당첨 통계( 당첨 내역 및 수익률 )를 출력한다.
  ```
  3개 일치 (5,000원) - x개
  4개 일치 (50,000원) - x개
  5개 일치 (1,500,000원) - x개
  5개 일치, 보너스 볼 일치 (30,000,000원) - x개
  6개 일치 (2,000,000,000원) - x개
  ```






docs/HISTORY.md
<div align="center">
  
# [기록] 로또 :slot_machine:	
<br>

본 문서는 우테코 6기 [**3주차 미션 - 자동차 경주**](https://github.com/woowacourse-precourse/javascript-lotto-6)의 구현 과정을 상세히 담고 있습니다.<br>
구현 기록은 `N차 구현`-`N차 구현 보충`의 형식으로 작성했습니다. :black_nib:

<br>

<div>
<img src="https://img.shields.io/badge/javascript-F7DF1E?style=for-the-badge&logo=javascript&logoColor=white"/>
<img src="https://img.shields.io/badge/node.js-339933?style=for-the-badge&logo=node.js&logoColor=white"/>
<img src="https://img.shields.io/badge/jest-C21325?style=for-the-badge&logo=jest&logoColor=white"/>
</div>

<br>

본 구현은 외부 라이브러리 없이 순수 Vanilla JS를 통해서만 구현했습니다.<br>
또한 `Node.js` 버전 `18.17.1` 이상의 실행환경이 필요합니다 😏

</div>

<br>

## :three:차 구현

```
🎯 각 기능에 대한 테스트 코드 작성에 집중한다.
```

<br>

기능 단위로 테스트를 진행한다. :two:차 구현에서 작성한 기능 목록

<br>

<div align="center">
   ⓐ 로또 구매<br>
   ⓑ 로또 발급<br>
   ⓒ 당첨 로또 생성<br>
   ⓓ 결과 계산<br>
</div>

<br>

을 단위로 총 네개의 테스트를 `tests > AdditionalTest.js`에서 진행한다.


<br>

## :two::arrows_counterclockwise:차 구현

```
🎯 2차 구현에 언급없이 추가적으로 구현된 부분에 대해 명시한다.
```

<br>

파일 및 코드 네이밍에 있어서도 자체적인 규칙을 두고 생성했다. :straight_ruler:

<br>

1. `utils` 디렉토리 내의 함수 및 파일 명은는 모두 `generate`를 접두어로 사용한다.

<br>

2. 각 model은 생성자에서 `validate` 함수를 통해 값을 검증한다. 이때 `validate` 함수 내에서 사용하는 모든 함수들은 `validator` 디렉토리 내의 `${model명}Validate` 파일내에 존재한다.

<br>

3. 상수는 모두 앞에 관련된 개념을 제시한다.

<br>

네이밍 컨벤션도 중요하지만, 무엇보다 중요한 것은 이름을 봤을 때 **어떤 역할을 하는지 들어나도록** 구체적으로 네이밍하는 것이다. :cat:

<br>

## :two:차 구현

```
🎯 디렉토리에 정확한 역할을 부여하고 데이터의 처리에 있어 패턴화한다. "일관적인 코딩"에 집중한다.
```

<br>

모든 데이터는 아래와 같은 과정을 거치도록 한다.

<div align="center">

**입력** ▷ **가공** ▷ **검증** ▷ **처리** ▷ **출력**

</div>

ⓐ 데이터를 입력받아 ⓑ 정수변환과 같은 가공을 거친 뒤, ⓒ 올바른 값인지 검증을하고 ⓓ 필요한 처리를 수행하여 ⓔ 출력한다.

이러한 패턴을 하나의 기능:sparkles:으로 생각하여 **일관성**있게 구현한다. 또한 각 과정에 대한 코드의 디렉토리를 분리한다. ⓐ는 `view`, ⓑ는 `util`, ⓒ는 `validate`, ⓓ는 `model`, ⓔ는 `view` 디렉토리에서 처리한다.

<br>

### :pencil2: 기능 목록

1. **로또 구매**

   ✦ 금액의 입력 ▷ 정수 변환 ▷ 1000으로 나누어 떨어지는 정수 형태인지 확인 ▷ 로또 구매 ▷ 구매한 로또의 개수 출력

<br>   

2. **로또 발급**

   ✦ 로또 개수 입력 ▷ 로또 번호 생성 ▷ 올바른 로또인지 확인 (길이, 중복) ▷ 로또 집합 생성 ▷ 로또 집합 출력

<br>

3. **당첨 로또 생성**

   ✦ 로또 번호 입력 ▷ 정수 배열로 변환 ▷ 올바른 로또인지 확인<br>
   ✦ 보너스 번호 입력 ▷ 정수로 변환 ▷ 올바른 보너스번호인지 확인 (로또 번호와의 중복)

<br>

4. **결과 계산**

   ✦ 당첨 번호와 로또 번호 비교 ▷ 점수판 생성 ▷ 점수판 출력

<br>

### :dizzy_face: 고민해야할 점

1. **Validation의 위치?**

   Validation은 효율상 가공의 과정에서 함께 진행하는 것이 좋다. 그러나 직관적인 코드를 위해서는 `model`의 생성자에서 진행하는 것이 가장 좋다.

<br>

2. **Set vs Array?**

   로또의 길이가 길어질 수록 Set을 이용하는 것이 좋아진다. 그러나 여전히 Array를 Set으로 변환하는 과정이 필요하다.

<br>

## :one:차 구현

```
🎯 1차 구현은 "MVC 패턴을 적용한 기능 구현"과 "테스트의 성공"에 집중한다.
```

<br>

* **Model**

   * 하나의 로또를 의미하는 객체 :slot_machine:<br>
      ✦ 로또간의 비교하는 기능을 포함한다.

   * 여러 로또의 집합을 정의하는 객체<br>
      ✦ 당첨번호와 로또 집합을 비교하여 결과를 산출한다.

<br>

* **Utils**

   * 유효한 로또를 생성하는 기능<br>
   ✦ 1부터 45까지의 숫자 중 6개의 번호와 1개의 보너스 번호를 뽑는다.<br>
   ✦ 번호들은 중복되어서는 안된다.

<br>

* **View**

   * 입력<br>
   ✦ 금액을 입력받는다. :money_mouth_face:<br>
   ✦ 당첨번호를 입력받는다.<br>
   ✦ 입력이 유효한지 확인한다. :white_check_mark:
   
   <br>

   * 출력<br>
   ✦ 게임 진행 문구를 출력한다.<br>
   ✦ 당첨 결과 및 수익률을 출력한다.

<br>

* **Controller**

   ✦ 게임의 입력부터 출력에 이어지는 흐름을 관리한다.

src/DB.js
import Lotto from './Lotto.js';

const STAGES = Object.freeze({
  NUM_1: 1,
  NUM_2: 2,
  NUM_3: 3,
  NUM_4: 4,
  NUM_5: 5,
  NUM_6: 6,
  EXIT: 7,
});

const RESULT= Object.freeze({
  NUM_3: [3, 0],
  NUM_4: [4, 1],
  NUM_5: [5, 2],
  NUM_5_BONUS: [7, 3],
  NUM_6: [6, 4],
});

const REWARD = [ 5000, 50000 , 1500000, 30000000, 2000000000];

const TEXTVIEW = Object.freeze({
  REQUEST_MONEYINPUT: '구입금액을 입력해 주세요.\n',
  LOTTO_AMOUNT: '개를 구매했습니다.',
  REQUEST_LOTTOWIN: '당첨 번호를 입력해주세요.\n',
  REQUEST_LOTTOBONUS: '보너스 번호를 입력해주세요.\n',
  RESULT: ([win3, win4, win5, win5Bonus, win6, percent]) => `3개 일치 (5,000원) - ${win3}개
4개 일치 (50,000원) - ${win4}개
5개 일치 (1,500,000원) - ${win5}개
5개 일치, 보너스 볼 일치 (30,000,000원) - ${win5Bonus}개
6개 일치 (2,000,000,000원) - ${win6}개
총 수익률은 ${percent}%입니다.`,
});

const CONST_VALUE = Object.freeze({
  LOTTO_PRICE: 1000,
  LOTTO_LENGTH: 6,
  LOTTO_RANGE_START: 1,
  LOTTO_RANGE_END: 45,
});

const INITIAL_VALUES = Object.freeze({
  NUM: 0,
  ARR: [],
  OBJECT: {},
  RESULT_ARR: [0,0,0,0,0,0],
});



class Data {
  static instance;

  #lottoBuy = INITIAL_VALUES.ARR;

  #lottoWin = INITIAL_VALUES.OBJECT;

  #lottoBonus = INITIAL_VALUES.NUM;

  #lottoCnt = INITIAL_VALUES.NUM;

  #moneyInput = INITIAL_VALUES.NUM;

  #result = INITIAL_VALUES.RESULT_ARR;

  resetData() {
    this.#moneyInput = INITIAL_VALUES.NUM;
    this.#lottoBonus = INITIAL_VALUES.NUM;
    this.#lottoBuy = INITIAL_VALUES.ARR;
    this.#lottoCnt = INITIAL_VALUES.NUM;
    this.#lottoWin = INITIAL_VALUES.OBJECT;
  }

  get result() {
    return this.#result;
  }

  set result(value) {
    this.#result = value;
  }

  resultAdd(idx){
    if(idx === null)return;
    this.#result[idx] += 1;
  }

  resultPercent(){
    let sum = 0;
    for(let i = 0; i < this.#result.length - 1; i += 1){
      sum += this.#result[i] * REWARD[i];
    }
    sum = sum * 100 / this.#moneyInput;
  
    let formattedSum = sum.toFixed(1);
    

    formattedSum = formattedSum.replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    
    this.#result[5] = formattedSum;
  }
  

  get moneyInput() {
    return this.#moneyInput;
  }

  set moneyInput(value) {
    this.#moneyInput = value;
  }

  get lottoBonus() {
    return this.#lottoBonus;
  }

  set lottoBonus(value) {
    this.#lottoBonus = value;
  }

  get lottoBuy() {
    return this.#lottoBuy;
  }

  set lottoBuy(value) {
    this.#lottoBuy = value;
  }

  addLottoBuy(value) {
    const addLotto = new Lotto(value);
    this.#lottoBuy = [...this.#lottoBuy, addLotto];
  }

  textLottoBuy() {
    let text = '';
    this.#lottoBuy.forEach((lotto) => {
      text = `${text}[${lotto.getNumbers()}]\n`;
    });
    return text;
  }

  get lottoCnt() {
    return this.#lottoCnt;
  }

  set lottoCnt(value) {
    this.#lottoCnt = value;
  }

  get lottoWin() {
    return this.#lottoWin;
  }

  set lottoWin(value) {
    this.#lottoWin = value;
  }
}

const db = new Data();

export {
  TEXTVIEW, db as Data, STAGES, INITIAL_VALUES, CONST_VALUE, RESULT,
};
